sematic/runners/state_machine_runner.py

```diff
@@ -253,9 +264,25 @@ def _get_seconds_to_next_timeout(
 
     def _cancel_non_terminal_futures(self):
         for future in self._futures:
-            if not future.state.is_terminal():
+            state = self._read_refreshed_state(future)
+            if state != future.state:
+                logger.warning(
+                    "In-memory future state '%s' differs from refreshed state"
```

Comment:
Is adding this extra call to the server for each future worth it for logging this warning?
Is this log actionable for developers or for users?

Reply:
It's not just for logging the warning--if they actually do differ, then we want to use the state from the server and not the local state. Besides, this code only runs on failed/canceled cleanup and thus an extra call doesn't matter too much.

Reply:
Seems kinda spammy:
```
[...]
  File "/app/sematic/examples/mnist/pytorch/mnist_train_default_image.binary.runfiles/sematic/sematic/runners/state_machine_runner.py", line 161, in _handle_sig_cancel
    raise CancellationError(
sematic.utils.exceptions.CancellationError: Pipeline run cancelled due to signal 15
2023-07-20 11:33:04,754 - INFO - sematic.resolvers.log_streamer: Cleaning up log ingestor
Showing the tail of the logs for reference. For complete logs, please use the UI. This contains the last file or two of log line deltas uploaded to remote storage.
		.
		.
		.
2023-07-20 11:33:02,968 - WARNING - sematic.runners.state_machine_runner: Received signal 15; canceling pipeline run...
2023-07-20 11:33:02,972 - WARNING - sematic.runners.local_runner: Received cancelation event
2023-07-20 11:33:02,972 - WARNING - sematic.runners.state_machine_runner: Received signal 2; canceling pipeline run...
2023-07-20 11:33:02,995 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.RAN' differs from refreshed state'FutureState.CANCELED' during cleanup.
2023-07-20 11:33:03,033 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.CREATED' differs from refreshed state'FutureState.CANCELED' during cleanup.
2023-07-20 11:33:03,055 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.CREATED' differs from refreshed state'FutureState.CANCELED' during cleanup.
2023-07-20 11:33:03,078 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.SCHEDULED' differs from refreshed state'FutureState.RESOLVED' during cleanup.
2023-07-20 11:33:03,102 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.CREATED' differs from refreshed state'FutureState.CANCELED' during cleanup.
2023-07-20 11:33:03,128 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.SCHEDULED' differs from refreshed state'FutureState.RESOLVED' during cleanup.
2023-07-20 11:33:03,226 - WARNING - sematic.utils.retry: Pipeline run cancelled due to signal 2
2023-07-20 11:33:03,227 - WARNING - sematic.utils.retry: Retrying request in 1 seconds with 6 tries left...
2023-07-20 11:33:04,248 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.RAN' differs from refreshed state'FutureState.CANCELED' during cleanup.
2023-07-20 11:33:04,271 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.CREATED' differs from refreshed state'FutureState.CANCELED' during cleanup.
2023-07-20 11:33:04,291 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.CREATED' differs from refreshed state'FutureState.CANCELED' during cleanup.
2023-07-20 11:33:04,311 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.SCHEDULED' differs from refreshed state'FutureState.RESOLVED' during cleanup.
2023-07-20 11:33:04,330 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.CREATED' differs from refreshed state'FutureState.CANCELED' during cleanup.
2023-07-20 11:33:04,349 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.SCHEDULED' differs from refreshed state'FutureState.RESOLVED' during cleanup.
2023-07-20 11:33:04,443 - WARNING - sematic.runners.state_machine_runner: Terminating runner loop due to cancellation.
2023-07-20 11:33:04,461 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.RAN' differs from refreshed state'FutureState.CANCELED' during cleanup.
2023-07-20 11:33:04,482 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.CREATED' differs from refreshed state'FutureState.CANCELED' during cleanup.
2023-07-20 11:33:04,503 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.CREATED' differs from refreshed state'FutureState.CANCELED' during cleanup.
2023-07-20 11:33:04,530 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.SCHEDULED' differs from refreshed state'FutureState.RESOLVED' during cleanup.
2023-07-20 11:33:04,554 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.CREATED' differs from refreshed state'FutureState.CANCELED' during cleanup.
2023-07-20 11:33:04,618 - WARNING - sematic.runners.state_machine_runner: In-memory future state 'FutureState.SCHEDULED' differs from refreshed state'FutureState.RESOLVED' during cleanup.
2023-07-20 11:33:04,720 - ERROR - __main__: Run failed: CancellationError: Pipeline run cancelled due to signal 15
[...]
```